lib/df-detection.c

```diff
@@ -0,0 +1,116 @@
+#include "linux/df-detection.h"
+#include <linux/context_tracking.h>
+#include <linux/entry-common.h>
+#include <linux/kernel.h>
+#include <linux/livepatch.h>
+#include <linux/sched.h>
+#include <linux/slab.h>
+#include <linux/stacktrace.h>
+
+void add_address(const void *addr, unsigned long len)
+{
+	if (current->addresses == NULL || current->pairs == NULL)
+		return;
+	if (current->num_read >= current->sz &&
+	    reallocate_extra_memory(current->sz, DF_MAX_RECORDS)) {
+		struct df_address_range *temp =
+		    (struct df_address_range *)krealloc(
+			current->addresses,
+			current->sz * 2 * sizeof(struct df_address_range),
+			GFP_KERNEL);
+		current->addresses = temp ? temp : current->addresses;
+		current->sz =
+		    current->addresses ? current->sz * 2 : current->sz;
+	}
+	if (current->num_read < current->sz) {
+		current->addresses[current->num_read].start_address = addr;
+		current->addresses[current->num_read].len = len;
+		detect_intersection();
+		current->num_read++;
+	}
+}
+
+void start_system_call()
```

Comment:
I think it makes sense to pass the syscall number to this function, otherwise we actually don't have any information about which syscall we are looking at.

Reply:
yeah that what i did